### DISCUSSION TITLE
TileDB support

### DISCUSSION THREAD

**Discussion by d-chambers** [DASDAE/dascore#155] [Ideas]
We (@d-chambers, @eileenrmartin, @jinwar)  had a good discussion with @chad-earthscope and Henry Berglund (couldn't find his github handle) about supporting [TileDB](https://tiledb.com/) in DASCore, specifically, the [earthscope schema](https://gitlab.com/earthscope/geocrate-group/timeseries) which they will use to store DAS data in AWS. This discussion is a place to keep track of details related to that implementation.

At first glance, it seemed the [DASCore machinery for adding file support](https://dascore.org/contributing/new_format.html) would work fine for TileDB. However, thinking more about this, it actually wont because DASCore assumes data are stored in files and directories have no semantic meaning beyond containing files. This is not the case with TileDB in which each instance *is a directory* with many different files.

In order to address this, I am thinking of having plugin support for custom spool types. This would allow us to have different types of spools returned `dascore.spool` based on the input type (which again, can be defined in external packages). As long as all the spools implement the `BaseSpool` interface this shouldn't be an issue for usability. It would also open support for passing in `Client`-like objects which connect to remote data sources. In the long run, we probably want these in a separate package anyway since testing them is always finicky. 

The only downside to this approach is that a directory of TileDB directories won't work with the `DirectorySpool`. I am not sure how important that is. I will think more about it and update this post later. Anyone else with ideas feel free to chime in.

**Comment by mbriggs134** [comment]
Hey dascore team,

I am a developer working for Earthscope and have started looking into the integration described above.  Some questions and I ideas have come to mind:

Questions:
1.  Does dascore currently support or plan to support paths like s3 URIs e.g. “s3://bucket/prefix/object”?
2.  For the plugin concept you are describing would I be developing all the code in a different repo which depends dasdae or is there a different model you had in mind?
3.  Also in addition to the custom spool type mentioned.  I assume I need to still add the 4 IO methods for generating patches from tileDB arrays?  Would this also be in a separate repo?

Ideas:
1.  Another way to support formats which are a collection of files like tileDB would be build the smarts into spool_from_str https://github.com/DASDAE/dascore/blob/8ebe6e6f1727f7ae1c9bb06793e4e9f5e351e3b7/dascore/core/spool.py#L364 by allowing formats to specify something like a get_path_format method.  This would allow in tileDB’s case a simple proxy to their method [tiledb.array_exists](https://tiledb-inc-tiledb.readthedocs-hosted.com/projects/tiledb-py/en/stable/python-api.html#tiledb.array_exists)
2. Along the lines of idea 1 this might also be a good place to think about adding s3 URI support

Just getting started becoming familiar with the library so any suggestions and thoughts are much appreciated!

**Comment by d-chambers** [comment]
Hi @mbriggs134, thanks for the comments, we look forward to working with you! 

Let me address some of your specific questions/comments by highlighting our current plan to support non-local resources and its reasoning. From my experience with ObsPy, the clients tend to be the most fickle part of the system, often causing test failures when services change/become unavailable. They are also very tricky to test locally unless the remote sever/environment is trivial to setup or there is a reliable, publicly-facing  service which doesn't require credentials. On top of that, they may require somewhat heavy dependencies, like Boto3, HTTPX, etc. Since a common use case is that a user collects several TB of DAS data then processes them on a workstation or small HPC cluster, we don't want DASCore inundated with the complexity associated with accessing remote resources. For those reasons, we are planning to create another library called DASClients (or something like that) in which we will add support for accessing various remote resources, potentially including S3, TileDB, Globus, etc.

DASClients would include:

- A way to manage credentials and create authenticated clients based on whatever service (e.g., AWS) 
- Spool implementations to interact with each resource (e.g., TileDB spool)

Onto your specific questions/comments:

> Does dascore currently support or plan to support paths like s3 URIs e.g. “s3://bucket/prefix/object”?

Yes, it would be part of DASClients

> For the plugin concept you are describing would I be developing all the code in a different repo which depends dasdae or is there a different model you had in mind?

Yes, we will just create DASClients in the dasdae org. DASClients will have DASCore as a dependency. I will send you, and whoever else works on it from Earthscope, an invite to the DASDAE org with the commit bit for DASClients once we have created it.

> In addition to the custom spool type mentioned. I assume I need to still add the 4 IO methods for generating patches from tileDB arrays?

Not immediately, we just need a spool implementation. The 4 methods you mention are for adding **file** support. From my admittedly limited understanding of TileDB, TileDB arrays can be stored in directories on a file system (as well as S3 blobs). Currently, the indexing machinery doesn't support directories that represent spools, only files.  This means we can't have nested TileDB folders, or TileDB folders insider folders with other DAS data. I have an idea how to fix this, but it might slow down our file indexing so I don't want to do it until we are sure we need it.


> Another way to support formats which are a collection of files like tileDB would be build the smarts into spool_from_str

Yes, this is a good idea. If all the logic for S3/TileDB is in DASClients though, we need to implement a spool-fetching plug-in system like we did for file-readers [here](https://dascore.org/contributing/new_format.html#register-plugin). This would allow DASClient to inject its spool discovery smarts into dascore. It's not yet clear to me what the interface should look like, and probably merits more discussion and thought.

Another idea might be to implement a thin wrapper around an S3 address which has a `to_spool` method. Then `dascore.spool` could have a shortcut for any object which has this method to just call it and return the result? So it might look like this:

```python
import dasclient
import dascore as dc

s3_blob = dasclient.S3_Client("s3://bucket/prefix/object”, boto_client=...) 


# spool and s3_spool would be the same. 
s3_spool = s3_blob.to_spool()
spool = dc.spool(s3_blob)
```

Any ideas you have are appreciated!

> Just getting started becoming familiar with the library so any suggestions and thoughts are much appreciated

Two things here:

First, we are overdue for a release. We are just trying to finish up a few important features, but I hope it will happen around the end of next week. It is probably best to wait until then to dive deep into DASCore because some of the internals have changed significantly. 

Second, we have bi-weekly developer meetings. If you have the chance to sit in on one it would be great to discuss these topics in more detail and you can meet the rest of the team. We are currently trying to find a new time to meet for the fall but we will make an announcement once those are setup.

**Comment by d-chambers** [comment]
I spent some more time thinking about this, and I think the right approach is to enable the `FiberIO` subclasses to be part of the `dascore.spool` dispatch. This would allow them to support more than just files, and avoids the need to implement any more plug-in machinery. 

So it might look something like this:

```python
from dascore.io import FiberIO

Class TileDBIO(FiberIO):
    """IO support for TileDB"""

    # specify supported input types here
    _supported_types = (str, Path, S3Client, ...)
```

Then `dascore.spool` can use the supported type information to dispatch whatever objects it receives to the appropriate `FiberIO` subclasses, which can be implemented in DASCore or separate packages. 

We could also add a `_supports_directories` (or something like that) to indicate directory paths can be provided and `_strs_are_paths` so a string which is not an accessible path on the file system doesn't get routed to `FiberIO` classes that expect such. 

What do you think?

**Reply by mbriggs134** [reply]
I like this idea of _supported_types allowing each plugin to be part of the `dascore.spool` dispatch.

I am a little unsure on the need for _supports_directories and _strs_are_paths concept.  Thinking about it briefly, I would hope the dispatcher would be able to determined a str to be something concrete, e.g. either a Path or S3Path, which would contain enough to determine if I can handle given "path".  If a given formats expects local file system then S3Path would not be registered.  I guess with the Path case requiring plugins to call Path.is_dir might be inefficient.  What are the cases that the dispatcher would pass along a str?

**Reply by d-chambers** [reply]
> I am a little unsure on the need for _supports_directories and _strs_are_paths concept

Right, I think we are on the same page but just to clarify: The semantic meaning of a string is ambiguous; it could be a local file, local directory, s3 path, url,  some kind of special keyword or alias to some external service, etc. If the spool dispatcher does some of the work in sorting what these strings mean and only provides the right type strings to the `FiberIO` instances, that simplifies the `FiberIO` implementation in an external package and means each FiberIO doesn't have to perform these checks, which probably results in net performance gains. 

For example, if there is a way to declare a `FiberIO` subclass only wants s3 paths (perhaps through a matching regex) then the dispatcher will simply not providing any strings that don't meet those conditions, and each `FiberIO` won't have to check itself.

This will take a bit more thought on how exactly it will work, but I think it is a promising direction.

**Reply by mbriggs134** [reply]
I agree we are on the same page.

One idea to implement the supported types is to use the scheme portion of urlparse e.g.
```
import os
from urllib.parse import urlparse
from pathlib import Path

rel_path = Path("write_example.tdb/")
abs_path = Path(os.path.abspath("write_example.tdb/"))
s3_uri = "s3://bucket/write_example.tdb/"

urlparse(rel_path.as_uri())
"""
ValueError: relative path can't be expressed as a file URI
"""

urlparse(abs_path.as_uri())
"""
ParseResult(scheme='file', netloc='', path='/Users/mbriggs/.../write_example.tdb', params='', query='', fragment='')
"""

urlparse(s3_uri)
"""
ParseResult(scheme='s3', netloc='bucket', path='/write_example.tdb/', params='', query='', fragment='')
"""
```
The initial obvious holes with this idea are it does not answer the directory versus file question and pushes us toward absolute paths which might cause other issues.

**Comment by d-chambers** [comment]
Hey @mbriggs134,

Sorry for the slow response on this. I just created [dasclients](https://github.com/DASDAE/dasclients) and added you as a maintainer. I put some skeleton setup for packaging, linting, and testing there but no real implementation yet. 

I will try to find some time to implement the dispatch ideas in this PR into DASCore sometime in the next few weeks so we can implement tiledb `FiberIO`. In the meantime, feel free to start working on the `TileDBSpool` implementation (which the `FiberIO`'s read method should return). Since it sub-classes the abstract base class `dascore.core.spool.BaseSpool` the interface is well-defined. You should see the methods you need to implement and their required signatures. Feel free to open a discussion over there if you have questions or need help.

**Comment by mbriggs134** [comment]
@d-chambers,

No worries on the response I know you are busy with your current release.  I will take a look at dasclients and hit you up with any questions / comments.  Just an FYI I will be taking a couple weeks off here so I might be a little slow on responding.

**Reply by d-chambers** [reply]
No worries at all. In fact, it might be best to pick this up when you return. That will give us some time to hopefully finish up the features for the release.
